examples/dreamcast/network/badsock/main.c

```diff
@@ -0,0 +1,350 @@
+        buf[rt - 2] = 0;
+
+    return 0;
+}
+
+int read_headers(http_state_t * hs, char * buffer, int bufsize) {
+    char fn[256];
+    int i, j;
+
+    for(i = 0; ; i++) {
+        if(readline(hs->socket, buffer, bufsize) < 0) {
+            if(i > 0)
+                return 0;
+            else
+                return -1;
+        }
+
+        if(strlen(buffer) == 0)
+            break;
+        
+        if(i == 0) {
+            if(!strncmp(buffer, "GET", 3)) {
+                printf("httpd: read GET header: '%s'\n", buffer);
+                for(j = 4; buffer[j] && buffer[j] != 32 && j < 256; j++) {
+                    fn[j - 4] = buffer[j];
+                }
+                
+                fn[j - 4] = 0;
+            }
+        }
+    }
+
+    strcpy(buffer, fn);
+    
+    return 0;
+}
+
+/**********************************************************************/
+
+int send_ok(http_state_t * hs, const char * ct) {
+    char buffer[512];
+
+    sprintf(buffer, "HTTP/1.0 200 OK\r\nContent-type: %s\r\nConnection: close\r\n\r\n", ct);
+    write(hs->socket, buffer, strlen(buffer));
+
+    return 0;
+}
+/* From here on things differ from the httpd example */
+/**********************************************************************/
+
+int do_dirlist(const char * name, http_state_t * hs) {
+    char * dl, *dlout;
+    int dlsize, r;
+
+    dl = malloc(65536);
+    dlout = dl;
+
+    sprintf(dlout, "<html><head><title>Listing of %s</title></head></html>\n<body bgcolor=\"white\">\n", name);
+    dlout += strlen(dlout);
+
+    sprintf(dlout, "<tr><td><a href=\"500.txt\">Click to download a 500B file.</a></td></tr>\n");
+    dlout += strlen(dlout);
+    sprintf(dlout, "<tr><td><a href=\"10000.txt\">Click to download a 10KB file.</a></td></tr>\n");
+    dlout += strlen(dlout);
+
+    dlsize = strlen(dl);
+    dlout = dl;
+
+    send_ok(hs, "text/html");
+    
+
+    while(dlsize > 0) {
+        r = write(hs->socket, dlout, dlsize);
+
+        if(r <= 0)
+            return -1;
+
+        dlsize -= r;
+        dlout += r;
+    }
+
+    free(dl);
+
+    return 0;
+}
+
+/**********************************************************************/
+
+#define BUFSIZE (256*1024)
+
+void *client_thread(void *p) {
+    http_state_t * hs = (http_state_t *)p;
+    char * buf;    
+    const char * sampletext = kos_get_banner();
+
+    printf("httpd: client thread started, sock %d\n", hs->socket);
+
+    buf = calloc(BUFSIZE, 1);
+
+    if(read_headers(hs, buf, BUFSIZE) < 0) {
+        goto out;
+    }
+
+    printf("httpd: client requested '%s'\n", buf);
+
+    /* Check for invalid or just '/' */
+    if(strlen(buf) <=1) {
+        do_dirlist(buf, hs);
+    }
+    else {
+        /* Starting at the second character, try to read as a dec number and put the end to leftover */
+        unsigned int size = strtol(buf + 1, NULL, 10);
+        
+        /* Couldn't pull a number */
+        if(size == 0) size=1000;
+
+        unsigned int o = size;
```

Comment:
```suggestion
       int o = size;
```
This needs to be an int because it will cause underflow issues with o -= r;

This still doesnt fix the infinite loop issue thought :(